pyhap/hap_server.py

```diff
@@ -190,9 +198,24 @@ def _upgrade_to_encrypted(self):
     def end_response(self, bytesdata, close_connection=False):
         """Combines adding a length header and actually sending the data."""
         self.send_header("Content-Length", len(bytesdata))
-        self.end_headers()
-        self.wfile.write(bytesdata)
-        self.close_connection = 1 if close_connection else 0
+        # Setting this head will take care of setting
+        # self.close_connection to the right value
+        self.send_header("Connection", ("close" if close_connection else "keep-alive"))
+        # Important: we need to send the headers and the
+        # content in a single write to avoid homekit
+        # on the client side stalling and making
+        # devices appear non-responsive.
+        #
+        # The below code does what end_headers does internally
+        # except it combines the headers and the content
+        # into a single write instead of two calls to
+        # self.wfile.write
+        #
+        # TODO: Is there a better way that doesn't involve
+        # touching _headers_buffer ?
+        #
+        self.connection.sendall(b"".join(self._headers_buffer) + b"\r\n" + bytesdata)
+        self._headers_buffer = []
```

Comment:
Is this really needed? I.e. how often do you see a stall if this is not done?

Reply:
I would see it drop the connection and then end up in a stall state for a while until iOS hit an internal logic to reconnect.  It would happen about once a week and I'd usually have to reboot the phone to make everything work again.  This fix is similar to an issue I've seen before with apple's webdav library.

Reply:
The underlying issue is likely a bug in apple's iOS libraries that isn't dealing well with data coming at unexpected times (reading when its expecting to write and vise-versa).  It is also likely the reason #215 also causes disconnects.  Apple developer support says its a "known" issue that is being worked on.

Reply:
Hey thanks for the insights, sounds good!